compiler/src/iree/compiler/Codegen/LLVMGPU/Utils/LLVMGPULayoutAnalysisAndDistribution.cpp

```diff
@@ -343,6 +344,49 @@ void propagateLayoutToFor(scf::ForOp forOp,
   }
 }
 
+static void propagateLayoutsToOthers(SmallVectorImpl<Value> &operands,
+                                     DenseMap<Value, Layout> &layoutMap) {
+  int numOperands = operands.size();
+  // Find an operand with a layout
+  int i;
+  for (i = 0; i < numOperands; i++) {
+    if (layoutMap.count(operands[i])) break;
+  }
+  // Propagate layout to others
+  for (int j = 0; j < numOperands; j++) {
+    if (j == i) continue;
+    if (!layoutMap.count(operands[j])) {
+      layoutMap.try_emplace(operands[j], layoutMap.at(operands[i]));
+      layoutMap.at(operands[j]).debugPrint("binary/unary operand");
+    } else {
+      assert(layoutMap.at(operands[i]) == layoutMap.at(operands[j]));
+    }
+  }
+}
+
+template <typename T>
+void propagateLayoutToBinaryOp(Operation *op,
+                               DenseMap<Value, Layout> &layoutMap) {
+  if (auto binaryOp = dyn_cast<T>(op)) {
+    Value lhs = binaryOp.getLhs();
+    Value rhs = binaryOp.getRhs();
+    Value result = binaryOp.getResult();
+    SmallVector<Value> operands{lhs, rhs, result};
+    propagateLayoutsToOthers(operands, layoutMap);
+  }
+}
+
+template <typename T>
+void propagateLayoutToUnaryOp(Operation *op,
+                              DenseMap<Value, Layout> &layoutMap) {
+  if (auto unaryOp = dyn_cast<T>(op)) {
+    Value operand = unaryOp.getOperand();
+    Value result = unaryOp.getResult();
+    SmallVector<Value> operands{operand, result};
+    propagateLayoutsToOthers(operands, layoutMap);
+  }
+}
```

Comment:
You should be able to generalize this by checking `hasElementwiseMappableTraits`. This way you can merge binary and unary and you don't need to make those templates.

Reply:
Even if you don't want to generalize to all elementwise at the time I would suggest just making a isElementwise helper and checking that to decide to treat the op as elementwise rather than using templates and having special cases for unary, binary etc...

Reply:
Makes sense. I can change this to use ```hasElementwiseMappableTraits```. This should definitely simplify the code.